### PR TITLE
feat(c4): migrate C4 relationships to the unified edge rendering

### PR DESCRIPTION
**Pull request by filipsajdak**
## What this is

The **edges slice** of the in-place C4 renderer migration (plan: #7849). **Standalone against `develop`.** (It was stacked on #7842, the shapes slice, which merged on 2026-07-30; the diff is now this PR's own 5 files, +202/-42.) Its commits are:

- `feat(c4): render C4 relationships through the unified edge rendering`
- `style(c4): dashed relationship lines and labels`
- `feat(c4): honour UpdateRelStyle offsets and colours for relationship labels`
- `fix(c4): place relationship endpoints via each shape's intersection`

## This slice: migrate relationships

The legacy `svgDraw` relationship drawing is replaced with mermaid's unified edge renderer (`insertEdge`/`insertEdgeLabel`): each `Rel` becomes a **dashed** edge with an arrowhead and an HTML label (`BiRel` -> arrowheads at both ends), matching c4model.com. The legacy grid still supplies the geometry (a straight line between the two shapes' boundary intersection points), fed to `insertEdge` with `skipIntersect`. Those endpoints now come from each shape's **own `node.intersect`** (its real outline), so arrows meet the person/cylinder/queue silhouette instead of a bounding box, with the rectangular approximation kept as a fallback. Boundaries and the grid **layout** are untouched - the layout slice is next.

`UpdateRelStyle` is honoured: `$offsetX`/`$offsetY` position the label (nudging it off nodes and separating parallel relationships) and `$textColor`/`$lineColor` recolour it. (The db stores named offsets as strings, so the renderer coerces them.)

## Impact / how to see it

No flag - the existing C4 cypress baselines now route relationships through the unified renderer, so **Argos shows the relationship change** per diagram. Layout/positions are unchanged from the shapes slice.

## Note

These `$offset` nudges are a legacy-syntax crutch; real auto-spacing arrives with the layout slice (dagre). Honouring them here keeps parity with the legacy renderer (no regression).

## Summary

Migrates C4 relationship rendering to Mermaid’s unified edge system as part of an in-place C4 renderer migration, replacing legacy `svgDraw` relationship drawing with `insertEdge` and `insertEdgeLabel`. Relationships now render as dashed edges with arrowheads (including `BiRel` arrowheads on both ends) and relationship labels are built/rendered through unified edge-label primitives. The legacy grid continues to supply relationship geometry and boundary intersection points, while layout/boundary recursion ordering is preserved via `async` rendering.

## Highlights

- **Unified C4 relationship rendering**
  - Switches C4 relationship drawing from `svgDraw` to `insertEdge` / `insertEdgeLabel`.
  - Dashed edge styling + marker/arrowhead insertion; `BiRel` supports arrowheads at both ends.
  - Relationship label rendering honors `UpdateRelStyle` for label offsets and color, including coercing stored string `$offsetX`/`$offsetY` values into numeric offsets when needed.
  - `drawC4ShapeArray`, `drawRels`, and boundary recursion are converted to `async` to ensure computed bounds/intersections exist before subsequent drawing steps.
  - Shape element DOM IDs are **namespaced per diagram ID** to prevent cross-diagram collisions.

- **New unified C4 shape/label adapters**
  - Adds **`c4ShapeAdapter.ts`** to adapt legacy C4 node/edge inputs into unified-renderer shapes/labels.
    - Resolves unified shapes via precedence: explicit `shape`/`sprite` → `tags` → legacy `typeC4Shape` mapping (including external variants) → default `rounded`.
    - Builds HTML-safe edge labels for relationships (escapes label text, and conditionally appends technology/description lines).
    - Constructs unified node objects via `buildC4Node` with computed CSS classes/styles and legacy layout fields retained for geometry.
  - Adds **`c4LabelHelper.ts`** for stacked C4-style labels as pure SVG text (name/type/stereotype + multi-line description), including wrapping behavior constrained by `c4.width`.
  - Updates the shared `labelHelper` to use `c4LabelHelper` when stereotyping is present.
  - Extends node types with optional `stereotype?: string` in `rendering-util/types.ts`.

- **C4 styling + edge/label CSS alignment**
  - Replaces minimal C4 `getStyles()` with a fuller C4 stylesheet in `src/diagrams/c4/styles.js`.
  - Adds per-C4-element label font rules generated from `getConfig().c4`.
  - Restyles C4 relationships to align with dashed unified edge rendering and updates marker/label styling (including background/opacity behavior for edge labels).

- **Person shape enablement**
  - Adds a **`person`** shape handler to the unified shapes registry (`rendering-util/rendering-elements/shapes.ts`).
  - Implements the **`person`** renderer (`person.ts`) using unified node/label helpers, including support for hand-drawn (`look === 'handDrawn'` / roughjs) rendering modes.
  - Adds a changeset for the new `person` shape and updates shape documentation snapshots accordingly.

- **Related shape/layout correctness tweaks**
  - Refactors `cylinder.ts` and `tiltedCylinder.ts` dimension normalization to improve width/height edge cases when only one dimension is provided.

- **Docs/tests/changesets updated**
  - Adds C4 documentation for **element text wrapping** (`wrap: true` + `c4.width`).
  - Updates Cypress C4 rendering characterization tests for element-shape overrides and description rendering/wrapping behavior.
  - Extends Cypress snapshot coverage with:
    - `C4.7`: per-element `personFontFamily`/`personFontSize` snapshot variations
    - `C4.8`: `c4.width` wrapping vs non-wrapping behavior
  - Updates `docs.spec.ts` inline snapshots by regenerating the shapes table after introducing C4 `c4-*` shapes and related shape docs.
  - Adds changesets (including `c4-unified-shapes`) and updates the cspell dictionary with `Structurizr`.

## Testing

- Full C4 unit suite passed, along with `multi-diagram-id-uniqueness` and `docs.spec`.
- Visual comparison against `develop` showed updated output across the C4 corpus tiles (including general/directional and `Rel_Back` relationship cases).
- Cypress E2E: an unrelated shard reported a Chrome renderer crash, but the remaining shard assertions were reported as passing.

### CODE REVIEW

**Comment by filipsajdak** [comment]
Both unit-test failures are fixed (the fixes live in the underlying #7842 commits, since both rooted in the shapes slice, and this branch carries them):

1. **`Multi-diagram ID uniqueness > "c4"`** - the unified shapes used the bare element alias (`user`, `system`) as the DOM id, so two identical C4 diagrams on one page collided. Fixed by namespacing the shape id per diagram (`node.domId = \`${diagramId}-${node.id}\``) in `drawC4ShapeArray`, matching how the other diagrams scope their node ids. Folded into `feat(c4): render C4 elements through the unified shapes`.
2. **`docs.mts > buildShapeDoc > should build shapesTable`** - registering the seven `c4-*` shapes added rows to the generated shapes table; the inline snapshot just needed regenerating (`vitest -u`). Folded into `feat(c4): add C4 element shapes to the unified shape registry`.

Both PRs were force-pushed (fixes squashed into the relevant commits, no fix-on-top noise); #7883 was rebased onto the new #7842 tip. Verified locally: full c4 unit suite (161 passing), `multi-diagram-id-uniqueness` and `docs.spec` both green.

**Comment by sidharthv96** [comment]
@filipsajdak please refrain from force pushing across 15+ PRs, as it blocks our CI for everyone else. We're also at 97% use for screenshot tests for this month, so just trying to be conservative. Thank you. 

image

**Comment by filipsajdak** [comment]
@sidharthv96 sorry, I was trying to keep the PRs in easy to merge state.

**Comment by sidharthv96** [comment]
> @sidharthv96 sorry, I was trying to keep the PRs in easy to merge state.

No worries, I was just letting you know of the problem. You've been doing amazing work with C4! 🚀 


It seems like the workflow of having multiple stacked PRs from a fork is poorly done in GitHub, as we cannot point to the previous branch as base.

**Comment by filipsajdak** [comment]
Re the CodeRabbit nitpicks: the `c4Renderer.ts` suggestions (edgeLabels map, knownBounds) target code the next slice **removes** - the layout migration deletes `c4Renderer.ts`/`svgDraw.ts` and moves C4 onto the unified pipeline (tracked in #7849), so investing there would be churn. The shape-boilerplate/reflow points are folded into the planned shared-shape-helper refactor after the migration slices merge.

**Comment by filipsajdak** [comment]
## Visual proof for this PR's actual diff (not the converged end-state)

Same context as [my note on #7842](https://github.com/mermaid-js/mermaid/pull/7842#issuecomment-4914064081): the #7849 tracker's gallery is a preview of the converged end-state combining several future phases (not any single open PR's current diff). This PR had no dedicated visual proof of its own either, so here is one via the same local `c4:visual-diff` harness (Argos-free): **`upstream/develop` (f3dea5838) vs this PR's branch tip (`ef9b917eb`)**.

Note: #7883 is stacked on #7842, so this diff includes #7842's shape changes too, plus #7883's own relationship/edge migration on top - that's expected, not double-counting.

25 of 25 corpus tiles changed (1.8%-14.5%). A few representative tiles, chosen for relationship coverage:

### `internet-banking-context` (14.51%) - whole-diagram overview

| before (develop) | after (this PR, stacked on #7842) | diff |
| --- | --- | --- |
| before | after | diff |

### `CHAR.rel-directions` (6.31%) - every directional `Rel`

| before | after | diff |
| --- | --- | --- |
| before | after | diff |

### `CHAR.rel-back` (4.72%) - `Rel_Back`

| before | after | diff |
| --- | --- | --- |
| before | after | diff |

Full 25-tile before/after/diff gallery + summary table: [summary.md](https://raw.githubusercontent.com/filipsajdak/mermaid/c4-proof-assets/pr-7883-2026-07-08/summary.md) (rendered HTML at [index.html](https://raw.githubusercontent.com/filipsajdak/mermaid/c4-proof-assets/pr-7883-2026-07-08/index.html), download to view locally since GitHub serves raw HTML as text).

**Scope reminder:** boundaries and the legacy grid layout are still untouched by this stack - that's the next slice (layout migration, opens once this lands).

**Comment by filipsajdak** [comment]
Hi @aloisklink - friendly nudge on this one too, since it's stacked on #7842. Same CI pattern: green aside from the expected `argos` visual-diff. No merge conflicts, no open threads. Whenever you have a moment, a review would be very welcome.

The full staged plan - what's landed, what's open, and the recommended merge order - is kept up to date in #7849.

Thanks for your time!

**Comment by filipsajdak** [comment]
CI note: the e2e (4) failure on the latest push is an unrelated flake - Cypress reports the Chrome renderer process crashed while starting rendering/iconShape-square.spec.ts (0 assertions failed; the other specs in the shard passed 332/332). I cannot re-run the job from a fork; a maintainer re-run of the failed job should bring e2e-required back to green.

File: packages/mermaid/src/diagrams/c4/c4ShapeAdapter.ts [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+const isExternal = (typeC4Shape: string): boolean => typeC4Shape.startsWith('external_');
+
+// Structurizr-style type line, e.g. `[Container: Node.js]`.
+const stereotypeText = (shape: C4ShapeLike): string => {
+  const stereotype = stereotypeLabel(shape.typeC4Shape.text);
+  return shape.techn?.text ? `[${stereotype}: ${shape.techn.text}]` : `[${stereotype}]`;
+};
+
+const escapeHtml = (txt: string): string =>
+  txt.replaceAll('&', '&amp;').replaceAll('<', '&lt;').replaceAll('>', '&gt;');
```
[the remark this answers is not in the document]
Comment (filipsajdak) [reply packages/mermaid/src/diagrams/c4/c4ShapeAdapter.ts]: Fixed in 5888d9503: `escapeHtml` now escapes double and single quotes as well.

File: packages/mermaid/src/rendering-util/rendering-elements/shapes/c4LabelHelper.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    .attr('style', handleUndefinedAttr(node.labelStyle));
+
+  const name = typeof node.label === 'string' ? node.label : (node.label?.[0] ?? '');
+  const sections = [
+    { text: name, cssClass: 'c4-name' },
+    { text: node.stereotype, cssClass: 'c4-type' },
+    ...(node.description ?? []).map((line) => ({ text: line, cssClass: 'c4-descr' })),
+  ].filter((section) => section.text);
+
+  const wrapWidth = Math.max((node.width ?? 0) - 2 * (node.padding ?? 0), MIN_WRAP_WIDTH);
```
[the remark this answers is not in the document]
Comment (filipsajdak) [reply packages/mermaid/src/rendering-util/rendering-elements/shapes/c4LabelHelper.ts]: Fixed in a6b7ea5d5 on the base PR (#7842) and merged through here in 60537c197: without `node.width` the helper now falls back to the flowchart `wrappingWidth` default instead of clamping to the 32px minimum.

**Comment by filipsajdak** [comment]
Re-pointed onto the new #7842 base. #7842 was approved and its history tidied to 9 commits (38 before), so this branch had to move with it; it was also `CONFLICTING` against `develop` for the same reason #7842 was (develop dropped `explicitDir` from `BaseNode`, colliding with `stereotype` in the generated `LayoutData.md`).

**History rebuilt rather than rebased: 17 commits -> 2.** A rebase was not available here - each original edge commit was authored against an early point of #7842, and this branch carried 8 merge-throughs of that branch, so the old patches no longer apply to the evolved files. I rebuilt forward from the final tree instead:

```
feat(c4): render C4 relationships through the unified edge rendering
style(c4): dashed relationship lines and labels
```

**The content is provably unchanged.** This PR touches exactly 3 files, and all three are byte-identical blobs to the previous tip `5f655c998`:

- `c4Renderer.ts`, `c4ShapeAdapter.ts`, `styles.js` - `git rev-parse 5f655c998:<file>` equals `git rev-parse HEAD:<file>` for each.
- The delta against the base is byte-identical too: diffing the old delta against the new one produces no output.

Also gone: the `fix(c4): place relationship endpoints via each shape's intersection` / `refactor(c4): drop the superseded endpoint-intersection helper` pair, which cancelled each other out. That work now lives in #7842 as a single commit, per @aloisklink's suggestion to move it there.

@sidharthv96 - your `Merge branch 'develop' into feature/c4-migrate-edges` (`632fad7c3`) is not in the new history. I checked before dropping it: it is the only non-authored commit in the range, it is a merge, and it carried no changes of its own beyond bringing `develop` in - which the new base now does directly. Happy to restore anything if I misread that. This is one force-push on this branch as part of the #7842 tidy, not a sweep across the stack.

Local checks on the new tip: c4 unit suite 126/126, `tsc` clean for the mermaid package. Still stacked on #7842, so this diff shrinks to just the relationship work once that merges.

**Comment by filipsajdak** [comment]
Reopened - this was closed in error a few minutes ago, not intentionally. Nothing about the PR changed; sorry for the noise on your notifications, @aloisklink @sidharthv96.

While I am here, the useful update: now that #7842 has merged, this PR stands on its own. It no longer carries the 20 shape-migration commits, and the diff is down from 23 files to **3**:

```
packages/mermaid/src/diagrams/c4/c4Renderer.ts     | 111 +++++++++++++--------
packages/mermaid/src/diagrams/c4/c4ShapeAdapter.ts |  20 ++++
packages/mermaid/src/diagrams/c4/styles.js         |  24 +++++
3 files changed, 113 insertions(+), 42 deletions(-)
```

That is the whole relationship migration: relationships render through the unified edge renderer (dashed lines, arrowheads, `BiRel` at both ends, `UpdateRelStyle` offsets and colours), with the legacy grid still supplying the straight-line geometry. c4 unit suite 126/126, `tsc` clean.

This is the last slice before the layout migration, which replaces the legacy `Bounds` grid with the unified dagre pipeline and deletes `c4Renderer.ts`/`svgDraw.ts` outright.

**Comment by filipsajdak** [comment]
Rebased onto current `develop` and dropped the trailing merge commit: **3 commits -> 2**, one per concern:

```
feat(c4): render C4 relationships through the unified edge rendering
style(c4): dashed relationship lines and labels
```

The tree is byte-identical to the previous head - the series was rebuilt from that exact tree rather than replayed, and `git diff <old head> <new head>` was empty before pushing. 126 c4 unit tests pass, `tsc` clean.

Still 3 files, +113/-42.

**Comment by filipsajdak** [comment]
This PR has not had a human review yet. Since #7842 merged it stands alone at 3 files, +113/-42, and it is the last dependency of the C4 layout slice.

One thing to know before opening the checks: the `argos` failure here is the intended visual change (relationships move to the unified edge renderer) sitting unapproved. Approving or rejecting those diffs in Argos *is* the visual review for this PR; there is no broken test behind it.

Full stack status: https://github.com/mermaid-js/mermaid/issues/7849#issuecomment-5222442429

**Comment by filipsajdak** [comment]
Thank you - and thank you especially for tracing the label pipeline end to end rather than eyeballing the escaping. That is a lot of work to do on someone else's PR.

All three important items and the first nit are addressed. One of them I have to correct, though the mistake behind it was mine.

## 🟡 The `getIntersectPoints` comment - corrected, and it is the comment that was wrong, not the code

The delegation you were looking for is there. `getIntersectPoint` is, in full:

```ts
const getIntersectPoint = function (fromNode: C4Shape, endPoint: Point): Point | null {
  if (!fromNode.intersect) {
    throw new Error(`C4 shape "${fromNode.alias}" has no intersect function. ...`);
  }
  const { x, y } = fromNode.intersect(endPoint);
  return new Point(x, y);
};
```

`c4Shape.intersect` is captured from the drawn unified shape in `drawC4ShapeArray` (`c4Renderer.ts:317`, `c4Shape.intersect = node.intersect;`), with the node pre-centred at its grid centre so the closure works in absolute coordinates. `C4Shape.intersect` is declared in `c4Types.ts:59`. That plumbing arrived with #7842 ("place relationship endpoints via each shape's intersection"), which is why a grep against a pre-#7842 base finds nothing - on this branch `grep -ri intersect packages/mermaid/src/diagrams/c4/` returns 15 hits. So arrows do meet the person and cylinder silhouettes, and `skipIntersect` is passed precisely because the points are already clipped.

But you were misled by my comment, and that is a fair thing to be misled by: it claimed a fallback to "the rectangular approximation" that does not exist - the guard **throws**. I have rewritten it to describe what the code does. A comment promising behaviour the code lacks is worse than no comment, and this one cost you real time. Sorry about that.

One genuine leftover you have surfaced indirectly: `getIntersectPoint`'s `Point | null` return type is now unreachable-null, which makes the caller's `if (!points.startPoint || ...)` guard dead. That arrived with #7842, so I have left it out of this PR rather than widen the diff - happy to clean it up in a follow-up if you would like it gone.

## 🟡 `htmlLabels: false` rendering markup as literal text - real regression, fixed

You are right, and the failure is exactly as you describe. With HTML labels off, `createText` takes the `nonMarkdownToLines` branch, which captures each tag as its own word and hard-codes `type: 'normal'` (`handle-markdown-text.ts:33-41`), so the tags reach `innerTspan.text(...)` and show as literal characters. Your repo's own test pins that behaviour at `handle-markdown-text.spec.ts:316-334`.

`buildEdgeLabel` now takes the effective setting and emits the plain form when HTML labels are off:

```
Uses<br/>[HTTPS]<br/>to read accounts
```

`<br/>` is kept in both forms deliberately - it is a line break in an HTML label and a line delimiter in a plain one, so the three-line shape survives either way.

I went with gating rather than `labelType: 'markdown'` because markdown gives bold and italic in both branches but has no equivalent for `<small>`, so the size distinction between the name and the technology/description lines would be lost for everyone. If you would rather have one code path than the styled default, say so and I will switch.

Covered by unit tests on `buildEdgeLabel`: both forms, the omitted-parts cases, and a hostile string (`<script>`, `&`, quotes) asserting it is escaped in the HTML form and left as-is in the plain one, where it is set as text content.

## 🟡 No changeset - added

`minor`, per the template's rule for features. Not deferring it to an umbrella changeset - one per PR is easier to audit, and #7842 already shipped with its own.

## 🟢 CSS injection hardening - taken

Colour values are now accepted only if they stand alone as a colour, via `CSS.supports('color', value)` with a conservative pattern fallback for environments without the CSS API (jsdom). A value like `red;background:url(...)` is dropped rather than interpolated. Cheap, and it keeps the surface no wider than the presentation-attribute approach it replaced.

## 🟢 The two remaining nits

- **Edge-label map retention** - agreed, and I have left it. C4 never calls `positionEdgeLabel`, so wiring up `clear()` for its benefit alone felt like reaching into the shared module for no behavioural gain; ids are unique per render. Happy to add it if you would prefer the hygiene.
- **`lastElementChild`** - agreed it is an assumption. It holds while C4 emits a mid-label only, and the comment says so. The alternative is a returned handle from `insertEdgeLabel`, which is a shared-signature change I would rather not make from this PR.

## Also

Both commit messages had `Co-Authored-By: Claude` / `Claude-Session:` trailers. Those are gone - I hit the same point on #7859 and it applies here equally, so I have not waited for you to raise it. The first commit message also described the label as unconditionally HTML, which is no longer true; corrected.

C4 unit tests: 131 pass (126 + the 5 new). `tsc` clean.

**Comment by filipsajdak** [comment]
Note on the red `lint` check: it is not a circular dependency, despite the workflow's error text.

The step failed while `npx` was installing madge itself:

```
npm error code ETARGET
npm error notarget No matching version found for @typescript-eslint/tsconfig-utils@8.67.0.
ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  mermaid@11.16.1 checkCircle: `npx madge --circular ./src`
```

`checkCircle` never ran, so the "Circular dependency detected" message is the workflow's generic failure text rather than a real finding. Running the same command on this branch locally:

```
$ npx madge --circular ./src
Processed 827 files (4.2s)
✔ No circular dependency found!
```

`lint` is green on the other open PRs and on `develop` (14:34 UTC today), so it looks like a transient registry resolution issue in that window. A re-run should clear it — I do not have permission to trigger one, and I would rather not push a filler commit given this repo does not squash-merge.

**Comment by filipsajdak** [comment]
Merged current `develop` in - #7874 landed and touches this same file, so this branch was conflicting.

One resolution worth recording, since it is a semantic merge rather than a textual one. Both sides changed `getIntersectPoints`:

- this branch rewrote it to aim each endpoint at the other shape's centre, so an arrow stops at the person or cylinder silhouette rather than at its bounding box;
- #7874 widened its parameters from `C4Shape` to `C4Element` so a boundary can be an endpoint.

Neither side's version is the answer on its own, so the widening is applied to the rewritten function. `drawRels`' `getC4ShapeObj` parameter is widened to match `c4Db.getC4Shape`, which #7874 changed to fall back to boundaries. The import hunk simply takes both sides.

132 C4 unit tests pass; tsc, eslint and Prettier clean. Merged rather than rebased, and the push is a fast-forward.

**Comment by filipsajdak** [comment]
Merged develop for the Cypress-to-Playwright migration. This branch adds no e2e specs, so the only conflict was the generated `LayoutData` typedoc page.

One note in case it recurs elsewhere: autofix.ci had pushed a regeneration of that page computed against the pre-merge base, so its line numbers disagreed with develop's post-merge ones. Develop's are current and autofix re-runs on the push. I did not regenerate locally - a local `docs:build` strips the GitHub source links, which would have been a much worse thing to commit.

132 C4 unit tests pass.

**Comment by filipsajdak** [comment]
Gentle nudge - no rush. The requested changes were addressed on 2026-08-20 and all review threads are resolved; the only red check is `argos` (7 changed snapshots) waiting on a baseline decision. The changed sheets are the relationship-rendering migration this PR is about, so they are expected rather than a regression.

**Comment by filipsajdak** [comment]
Same ask as #7859: could a maintainer **dismiss or refresh the `knsv-bot` review**? `reviewDecision` is still `CHANGES_REQUESTED` from 2026-08-10, and the commits addressing it landed 2026-08-20. This one matters more than most, because **#8042 is stacked on it** - the two together are what actually delete `c4Renderer.ts` and `svgDraw.ts` and finish the migration.

All three "things to address" are in, and each is checkable:

| review item | status |
| --- | --- |
| `getIntersectPoints` comment describes behaviour the code does not have | **rewritten** - the "real outline / silhouette rather than a bounding box" wording is gone; grep returns zero |
| `htmlLabels: false` renders markup as literal text | **fixed** - `buildEdgeLabel` takes the effective setting via `getEffectiveHtmlLabels` and emits a plain form when HTML labels are off, with unit tests for both branches and a hostile string |
| no changeset | **added** - `.changeset/c4-unified-edges.md` |

The `CSS.supports` colour-validation nit is in as well, which then tripped `unicorn/better-regex`; autofix.ci pushed the corrected pattern and I left that commit rather than spend a screenshot run folding it away.

Branch `01f3e5b99`, own diff 5 files (+202/-42). Threads all resolved; only red is `argos` (7 changed).

Also correcting the PR body: it still says "stacked on #7842", which merged on 2026-07-30. **This is standalone against `develop` now** - I will update the description.

**Comment by filipsajdak** [comment]
Correcting one line of my comment above, because it undersells what happened and a maintainer reading it would check the wrong thing.

I wrote that the `getIntersectPoints` comment was "rewritten" and that the old wording is gone. The wording is gone, but that is not how the finding was resolved. The review offered two options - *"either wire up the `node.intersect` lookup the comment promises, or correct the comment to match reality"* - and **the first one is what happened**: `c4Renderer.ts:340` captures `c4Shape.intersect = node.intersect` from the unified shape that drew each element, and `:366` uses it (`fromNode.intersect(endPoint)`). So arrows really do stop at the person/cylinder silhouette now, and the comment describes what the code does.

The PR description said the same thing all along and is therefore accurate; I have only corrected its stale opening line, which still claimed this was stacked on #7842 (merged 2026-07-30). It is standalone against `develop`.

Sorry for the double post - better to have the record right than tidy.
